### DISCUSSION TITLE
If spdlog is header-only then is there any need to link with the DLL?

### DISCUSSION THREAD

**Discussion by siddhsql** [gabime/spdlog#3180] [Q&A]
I am not very proficient with C++. From what I understand header-only means all the source code of the library is in header files. There are no cpp files. If we use any DLL in C++, we need to import the headers that tell the compiler what functions are exported by the DLL. But in this case if we import the headers we will get the source code as well. So is there any need to use the DLL at all? What difference does it make if any? 

In fact, shouldn't linking against a DLL create a problem since you now have two "versions" - one version as a result of compiling the header files and another version in the DLL?

**Comment by gabime** [comment]
You dont need the dll if you are using the header only version. The advantage of dll is that it reduces compile time of your code significantly since  spdlog doesn't get compiled again and again when included by your code.

**Reply by siddhsql** [reply]
Could you explain to me? if I use the dll I still have to include the header, and if I include the header it will get compiled. plus now i have a duplicate copy of every function - one in the dll that i am linking to and another in the header that got included. how does c++ handle this?

**Comment by siddhsql** [comment]
btw one of the reasons i am asking this question because on this page:
https://github.com/gabime/spdlog/wiki/How-to-use-spdlog-in-DLLs
it says

>Because spdlog is header-only, building shared libraries and using it in a
main program will not share the registry between them.

suggesting that if one uses the spdlog dll, this problem will not happen.
but if my understanding is correct, this problem will happen no matter
whether you use spdlog in header-only form or if you link against the dll.
